File: DataAlgo_forPy/Graphs/adjacency_list.js
class Vertex {
  constructor(key) {
    this.id = key
    this.connectedTo = {}
  }

  addNeighbor(neighbor, weight = 0) {
    this.connectedTo[neighbor] = weight
  }

  getConnections() {
    return Reflect.ownKeys(this.connectedTo)
  }

  weight(neighbor) {
    return this.connectedTo[neighbor]
  }

  toString() {
    return `${this.id} connected to: ${Reflect.ownKeys(this.connectedTo)}`
  }
}

class Graph {
  constructor() {
      this.vertList = {}
      this.numVerticies = 0
  }

  [Symbol.iterator]() {
      const vertListKeys = Reflect.ownKeys(this.vertList)
      let index = 0
      return {
        next: () => {
          let value = Reflect.get(this.vertList, vertListKeys[index])
          index++
          return {
            done: index > vertListKeys.length ? true : false,
            value: value
          }
        }
      }
    }

    // Iterate through vertList values
  each(callback) {
      for (let vertex of this) {
        callback(vertex)
      }
    }

  contains(vertex) {
    return Reflect.get(this.vertList, vertex)
  }

  addVertex(key) {
      const newVertex = new Vertex(key)
      this.vertList[key] = newVertex
      this.numVerticies++
          return newVertex
  }

  getVertex(vertex) {
    if (this.vertList.hasOwnProperty(vertex)) {
      return this.vertList[vertex]
    } else {
      return null
    }
  }

  addEdge(fromVertex, toVertex, cost = 0) {
    if (!this.vertList.hasOwnProperty(fromVertex)) {
       this.addVertex(fromVertex)
    }
    if (!this.vertList.hasOwnProperty(toVertex)) {
       this.addVertex(toVertex)
    }

    this.vertList[fromVertex].addNeighbor(this.vertList[toVertex], cost)
  }

  getVerticies() {
    return Reflect.ownKeys(this.vertList)
  }

}

const g = new Graph()

for (var i = 0; i < 6; i++) {
  g.addVertex(i)
}

g.addEdge(0, 1, 2)

g.each((vertex) => {
  console.log("vertex: ", vertex)
  console.log("Connections: ", vertex.getConnections())
})

File: DataAlgo_forPy/Graphs/adjacency_list.rb
class Vertex
  def initialize(key)
    @id = key
    @connectedTo = {}
  end

  def addNeighbor(neighbor, weight)
    @connectedTo[neighbor] = weight
  end

  def getConnections
    @connectedTo.keys
  end

  def weight(neighbor)
    @connectedTo[neighbor]
  end

  def toString
    "#{@id} is connected to #{getConnections}"
  end

end


class Graph
  def initialize
    @vertList = {}
    @numVerticies = 0
  end

  def each(&prc)
    @vertList.each do |identification, connections|
      prc.call(identification, connections)
    end
  end

  def contains?(target)
    @vertList.has_key?(target)
  end

  def addVertex(key)
    newVertex = Vertex.new(key)
      @vertList[key] = newVertex
      @numVerticies += 1
    newVertex
  end

  def getVertex(key)
    @vertList[key] || nil
  end

  def addEdge(from_vertex, to_vertex, cost)
    unless @vertList.has_key?(from_vertex)
      @vertList.addVertex(from_vertex)
    end

    unless @vertList.has_key?(to_vertex)
      @vertList.addVertex(to_vertex)
    end

    @vertList[from_vertex].addNeighbor(@vertList[to_vertex], cost)
  end

  def getVerticies
    @vertList.keys
  end


end


g = Graph.new

6.times do |i|
  g.addVertex(i)
end

g.addEdge(0, 1, 2)

g.each do |vertex, connections|
  p "#{vertex} has connections #{connections.toString}"
end
